File: src/ch/bailu/aat/preferences/SolidInteger.java
package ch.bailu.aat.preferences;

public class SolidInteger extends SolidType {
    private final String key;
    private final Storage storage;
    
    
    public SolidInteger(Storage s, String k) {
        storage=s;
        key=k;
    }

    
    public int getValue() {
        return getStorage().readInteger(getKey());
    }
    
    
    public void setValue(int v) {
        getStorage().writeInteger(getKey(),v);
    }
    

    @Override
    public String getKey() {
        return key;
    }

    @Override
    public Storage getStorage() {
        return storage;
    }
    
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/preferences/SolidInteger.kt
package ch.bailu.aat_lib.preferences

import ch.bailu.aat_lib.exception.ValidationException
import ch.bailu.aat_lib.logger.AppLog.e
import ch.bailu.aat_lib.resources.Res

open class SolidInteger(private val storage: StorageInterface, private val key: String) : AbsSolidType() {
    open fun getValue(): Int {
        return getStorage().readInteger(getKey())
    }

    open fun setValue(value: Int) {
        getStorage().writeInteger(getKey(), value)
    }

    override fun getKey(): String {
        return key
    }

    override fun getStorage(): StorageInterface {
        return storage
    }

    override fun getValueAsString(): String {
        return getValue().toString()
    }

    @Throws(ValidationException::class)
    override fun setValueFromString(string: String) {
        var s = string
        s = s.trim { it <= ' ' }
        if (!validate(s)) {
            throw ValidationException(String.format(Res.str().error_integer(), s))
        } else {
            try {
                setValue(s.toInt())
            } catch (e: NumberFormatException) {
                e(this, e)
            }
        }
    }

    override fun validate(s: String): Boolean {
        return s.matches(VALIDATE)
    }

    companion object {
        // only positive/negative Integers, any size allowed
        private val VALIDATE = Regex("-?[1-9]\\d*")
    }
}
